scripts/inference_fp8.py

```diff
@@ -0,0 +1,287 @@
+# Example usage of 13B model on 2 GPUs with Tensor Parallel:
+# srun -N 1 --gres=gpu:2 torchrun --nproc_per_node=2 scripts/inference.py --model_path=~/models/13B-F --tokenizer=~/models/tokenizer.model --distributed
+
+parser = argparse.ArgumentParser(
+    description="Script to run inference on a causal model"
+)
+parser.add_argument("--device_type", type=str, default="cuda")
+parser.add_argument(
+    "--architecture",
+    type=str,
+    default="llama",
+    help="The model architecture to benchmark",
+)
+parser.add_argument(
+    "--variant",
+    type=str,
+    default="7b",
+    help="The model variant (configuration) to benchmark. E.g. 7b, 13b, 70b.",
+)
+parser.add_argument(
+    "--model_path",
+    type=str,
+    help="Path to the directory containing LLaMa weights (.pth files sharded by tensor parallel rank, not HF weights)",
+)
+parser.add_argument(
+    "--model_source",
+    type=str,
+    help="Source of the checkpoint. E.g. 'meta', 'hf', None",
+)
+parser.add_argument(
+    "--tokenizer",
+    type=str,
+    required=True,
+    help="Path to the tokenizer (e.g. ~/tokenizer.model)",
+)
+parser.add_argument(
+    "--no_use_cache",
+    action="store_false",
+    help="Disable the kv-cache (on by default)",
+)
+parser.add_argument(
+    "--unfuse_weights",
+    action="store_true",
+    help="If set to True, this will unfuse any fused weight modules that support the unfuse_weights method",
+)
+parser.add_argument(
+    "--compile",
+    action="store_true",
+    help="Use torch.compile (slow for first inference pass)",
+)
+parser.add_argument(
+    "--compile_mode",
+    type=str,
+    help="Mode for compilation",
+    default="default",
+    choices=["default", "reduce-overhead"],
+)
+parser.add_argument(
+    "--deterministic",
+    action="store_true",
+    help="Set torch.use_deterministic_algorithms? Requires env variable `CUBLAS_WORKSPACE_CONFIG=:4096:8`",
+)
+parser.add_argument(
+    "--distributed",
+    action="store_true",
+    help="This is a distributed job (multiple instances run with RANK+WORLD_SIZE)",
+)
+parser.add_argument(
+    "--batch_input",
+    action="store_true",
+    help="use a batch of prompts as input",
+)
+parser.add_argument(
+    "--min_pad_length",
+    type=int,
+    help="Pad inputs to a minimum specified length. If any prompt is larger than the specified length, padding will be determined by the largest prompt",
+    default=0,
+)
+parser.add_argument("--context_file", type=str, default=None, help="File to summarize")
+parser.add_argument(
+    "--unfuse",
+    action="store_true",
+    help="instantiate unfused model instance`",
+)
+args = parser.parse_args()
+
+local_rank = int(os.getenv("LOCAL_RANK", 0))
+world_size = int(os.getenv("WORLD_SIZE", 1))
+if args.device_type == "cuda":
+    device = torch.device(args.device_type, local_rank)
+    torch.cuda.set_device(device)
+else:
+    device = torch.device(args.device_type)
+
+torch.set_default_dtype(torch.bfloat16)
```

Comment:
if we add a dtype arg to the original inference.py, can we just have some conditions on `if dtype == fp8` instead of having a whole new inference script?

Reply:
yeah, this script is for debugging convenience for now - it will not be included in the final PR